test/Helper/Service/DoctypeFactoryTest.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/**
+ * @see       https://github.com/laminas/laminas-view for the canonical source repository
+ * @copyright https://github.com/laminas/laminas-view/blob/master/COPYRIGHT.md
+ * @license   https://github.com/laminas/laminas-view/blob/master/LICENSE.md New BSD License
+ */
+
+namespace LaminasTest\View\Helper\Service;
+
+use Interop\Container\ContainerInterface;
+use Laminas\ServiceManager\ServiceManager;
+use Laminas\View\Helper\Doctype;
+use Laminas\View\Helper\Service\DoctypeFactory;
+use PHPUnit\Framework\TestCase;
+
+final class DoctypeFactoryTest extends TestCase
+{
+    protected function setUp(): void
+    {
+        Doctype::unsetDoctypeRegistry();
+    }
+
+    public function testServiceIsCreatedOnInvocation(): void
+    {
+        $container = $this->createContainer();
+
+        $factory = new DoctypeFactory();
+        $service = $factory($container, '');
+
+        self::assertInstanceOf(Doctype::class, $service);
+    }
+
+    public function testFactorySetsDoctypeBasedOnConfig(): void
+    {
+        $config    = ['view_helper_config' => ['doctype' => Doctype::XHTML1_STRICT]];
+        $container = $this->createContainer($config);
+
+        $factory = new DoctypeFactory();
+        $service = $factory($container, '');
+
+        self::assertSame(Doctype::XHTML1_STRICT, $service->getDoctype());
+    }
+
+    public function testDefaultDoctypeIsUsedIfConfigIsMissing(): void
+    {
+        $config    = ['view_helper_config' => []];
+        $container = $this->createContainer($config);
+
+        $factory = new DoctypeFactory();
+        $service = $factory($container, '');
+
+        self::assertSame(Doctype::HTML4_LOOSE, $service->getDoctype());
+    }
+
+    /**
+     * @param array<string, mixed> $config
+     * @return ContainerInterface
+     */
+    private function createContainer(array $config = [])
```

Comment:
> I almost never use mocks

That is your personal decision and you can do this in your projects. Thats totally fine if you are doing this.
However, in almost every laminas package, we are using mocks instead of concrete implementations as it solves multiple issues:
1. If a bug rises in `laminas-servicemanager`, this package is not affected directly
2. "Unit tests" are supposed to test a specific unit and not how they interact with their dependencies. Since there is no concrete dependency to the `servicemanager`, there is no need to use the servicemanager as a dependency
3. You are explicitly testing the behavior on how the unit is interacting with the container, without using a mock, you cannot verify if the config was called from the container of if it was assembled within the method you are calling

So, please change this to a mock as it would provide better maintainability of this package. 👍

Reply:
Thanks for the input! I changed the test to use a Mock. But I still don't really get why a Mock is better than a "real" object, but that's not something to be discussed here 🙂 Maybe sometime in Slack with a (virtual) beer 😉